src/Network/Affjax/DSL.purs

```diff
@@ -0,0 +1,101 @@
+module Network.Affjax.DSL
+  ( AffjaxRequest()
+  , AffjaxRequestF(..)
+  , affjaxRequest
+  , url
+  , method
+  , method'
+  , header
+  , header'
+  , content
+  , content'
+  , username
+  , username'
+  , password
+  , password'
+  ) where
+
+import Control.Monad.Free (FreeC(), liftFC, runFreeCM)
+import Control.Monad.State (State(), execState)
+import Control.Monad.State.Class (modify)
+import Data.Coyoneda (Natural())
+import Data.Maybe (Maybe(..))
+import Network.Affjax.Request
+import Network.HTTP (Verb(..), Header(..), HeaderHead())
+
+-- | A free monad for building AJAX requests
+type AffjaxRequest c = FreeC (AffjaxRequestF c)
+
+-- | The request DSL AST.
+data AffjaxRequestF c a
+  = SetURL String a
+  | SetMethod MethodName a
+  | AddHeader Header a
+  | SetContent (Maybe (Content c)) a
+  | SetUsername (Maybe String) a
+  | SetPassword (Maybe String) a
+
+-- | The interpreter for the request DSL AST.
+affjaxN :: forall c. Natural (AffjaxRequestF c) (State (AjaxRequest c))
+affjaxN (SetURL url a) = const a <$> modify (_ { url = url })
+affjaxN (SetMethod method a) = const a <$> modify (_ { method = method })
+affjaxN (AddHeader header a) = const a <$> modify (\req -> req { headers = header : req.headers })
+affjaxN (SetContent content a) = const a <$> modify (_ { content = content })
+affjaxN (SetUsername username a) = const a <$> modify (_ { username = username })
+affjaxN (SetPassword password a) = const a <$> modify (_ { password = password })
+
+-- | Runs the DSL, producing an `AjaxRequest` object.
+affjaxRequest :: forall c a. AffjaxRequest c a -> AjaxRequest c
+affjaxRequest = (`execState` defaultRequest) <<< runFreeCM affjaxN
+
+-- | Take a standard HTTP verb and allow it to be used as an AJAX request
+-- | method.
+methodNameFromVerb :: Verb -> MethodName
+methodNameFromVerb = MethodName <<< show
+
+-- | Sets the URL for a request.
+url :: forall c. String -> AffjaxRequest c Unit
+url url = liftFC (SetURL url unit)
```

Comment:
I'd also like to see combinators like `get "http://foo.com"` which set the method and the URL at the same time, so you could write code like:

```purescript
do get "http://foo.com"
   ...
```